### ISSUE TITLE
code is always inlined into html page

### ISSUE THREAD

**Issue by MartinNowak** [Nekroze/diet-duo#1]
This is pretty inefficient for layouts, because the css/js is delivered for every page.
There should be a workflow that allows to precompile assets and use those in the layout.

For example a index.css/index.js combined with a `duo public/index.css | uglifycss > public/index-$(md5sum public/index.css | cut -d' ' -f1).css` and a subsequent reference of `index-md5sum.css` would work.

**Comment by Nekroze** [comment]
Good idea, I should be able to make it output to file instead and return the html to pull it in. Memoization should take care of making sure it runs only once but I will still need to make sure it runs if the source was updated... I will look into this.

**Comment by Nekroze** [comment]
Honestly though this project is a failure and pointless if I cannot get it to generate the duo output at compiletime. As you have noted and I have confirmed in my testing, the templates are compiled at compile time but not evaluated or some such until runtime... which incurs a big long delay for the first page load while duo compiles the resources.

**Comment by MartinNowak** [comment]
It's still a pretty good first step. Something like a postBuildCommand might come in handy to precompile assets. Maybe you can add a asset_path function that can be used to include css/js like so.
```jade
link(rel='stylesheet', href='#{asset_path("css/index.css")}')
script(src='#{asset_path("js/index.is")}')
```
At runtime asset_path determines the hash of the resource and forwards to e.g. `css/index-71f920fa275127.css` which is the precompiled version of that file, generated by the post build command.
You could probably even provide a diet include with two functions.

`asset_pipeline.dt`
```jade
- import diet_duo : asset_path;
- void include_css(string path)
  link(rel='stylesheet', href='#{asset_path(path)}')
- void include_is(string path)
  script(src='#{asset_path(path)}')
```

Then it can be used like so.
```jade
include asset_pipeline

- include_js("js/index.is")
- include_css("css/index.css")
```

**Comment by Nekroze** [comment]
@MartinNowak 

From what I can tell the D compiler doesn't support executing commands (via `std.process` for instance) at compile time so the original concept will not work regardless. 

However, I may be able to move the generation of these resources to runtime and make it run duo in a unittest guard so this way it is only generated when doing a test, which users of *diet-duo* would do before packaging/distribution. Then the normal runtime code that interfaces with D/Vibed could use a method such as you discribed to include it.

This way the executable to be released, and the web server any software using *diet-duo*, would never have to have duo installed. Which was the original idea.

Thoughts?

**Comment by MartinNowak** [comment]
You can add a second executable configuration to the project that allows to use `dub run diet-duo`.

**Comment by MartinNowak** [comment]
Dub itself is a library/executable package. Have a look at https://github.com/D-Programming-Language/dub/blob/master/dub.json.

**Comment by Nekroze** [comment]
@MartinNowak 

I like that idea. But would it be as clean as putting the compile command in a unittest block. Because with that you get to pretty much test your javascript/css code at the same time as your D code.

Also with that idea how would the end user write a compile so it runs all compile's with that command? in a module level version block?

**Comment by MartinNowak** [comment]
I'd use the `postBuildCommands` to invoke `dub run diet-duo`.
http://code.dlang.org/package-format#build-settings

**Comment by Nekroze** [comment]
@MartinNowak 
I have removed diet-duo from dub in favor of the new project. It is not ready for dub release yet so is only only personal repository, but you can review the current dev branch if you would like at https://projects.nekroze.com/projects/duod and please do excuse the self signed certificate error.

Currently it is all slated to work with unittest for the duo build and yuglify minification phase and should be able to be called from within a diet template as well very much like the original duo `require` command. I will however, once stabilized and working, look at dub command integration.

Thanks a heap for your assitance.

**Comment by Nekroze** [comment]
And closed.

**Comment by MartinNowak** [comment]
Looks promising.
Using unittests for the actual generation is still fairly hacky. Sometimes you just want to write a small prototype and don't care about testing or you forgot a particular test and only notice when clicking on that page.
I'd focus on precompiling/including complete js/css files for now, and leave sugar like require in diet templates aside for now. We can probably improve diet's textfilter plugin mechanism so that it can render at compile time.

**Comment by MartinNowak** [comment]
I filed an enhancement request for vibe.d.
https://github.com/rejectedsoftware/vibe.d/issues/985

**Comment by MartinNowak** [comment]
Some more info on Rails' asset pipeline.
http://guides.rubyonrails.org/asset_pipeline.html

**Comment by Nekroze** [comment]
@MartinNowak 
Indeed utilising unittests is a bit hacky but I still think it makes sense to have a testing phase for your assets. I would like to provide some output to the duo compiling for unittest reporting as well. For example at the moment it reports the outputed assets size which you can track over time as you add or remove dependencies.

This could also be used to test programatic generation of the asset code that gets handed to duo for example.

While I do get that sometimes in a prototype you may not run a test but if you know thats how one of your modules works then you will ensure it does then.

What do you mean just focus on precompiling and including? that is what I have done here, the sugure to make `require` like functionality was just easy to add on the side for convenience.

However to address some of these concerns I can pull the compilation of assets out of the unittest version block in a way that, if the asset is entirely missing at runtime (for example a unittest (which would always build with duo) has not been run for it), it will try to build it at runtime on its first use as diet-duo did as a kind of fallback so it still works running on a dev machine with duo installed.

**Comment by MartinNowak** [comment]
> What do you mean just focus on precompiling and including? that is what I have done here, the sugure to make require like functionality was just easy to add on the side for convenience.

Focus on precompiling css/js files in the public folder. That's more important than embedding duo in diet. The latter is nice to have, but the text filter plugins aren't powerful enough yet to support this use-case.

> However to address some of these concerns I can pull the compilation of assets out of the unittest version block in a way that, if the asset is entirely missing at runtime (for example a unittest (which would always build with duo) has not been run for it), it will try to build it at runtime on its first use as diet-duo did as a kind of fallback so it still works running on a dev machine with duo installed.

Don't, that'll only cause confusion and hides the issue that the assets weren't precompiled.

> however due to not being able to execute external commands during CTFE it would not solve this specific issue as duo would need to be invoked

It'll work by registering each css/is block in a shared static this() method. Then you could precompile all templates using something like `./myapp --precompile-assets`.

**Comment by Nekroze** [comment]
>It'll work by registering each css/is block in a shared static this() method.

This would require the user to declare `static this()` blocks where they may not already have them. This seems much the same to me as the current method of using `unittest` blocks instead of `static this()` except it would require an additional step/command to be added to any given projects build steps.

>Focus on precompiling css/js files in the public folder. That's more important than embedding duo in diet.

I completely agree and the current focus and example usages of DuoD is just that, compiling asset files. While it does support input and output of D strings that is more just me seperating code into logical blocks/functions.

**Comment by MartinNowak** [comment]
> This would require the user to declare static this() blocks

No you can generate them in the template, so it requires https://github.com/rejectedsoftware/vibe.d/issues/985.

**Comment by Nekroze** [comment]
@MartinNowak 
Ah I see, a `static this` in a D code block in a diet template... interesting.

**Comment by MartinNowak** [comment]
I recently used this [here](http://dpaste.dzfl.pl/cd375ac594cf#line-15) to solve [this](http://forum.dlang.org/post/eeywxusjhcapltfswqxz@forum.dlang.org) problem.
Given that we improve diet, which isn't that hard, you can globally register every use of a duo textfilter without running them. Then it's easy to start the app, compile all collected scripts and exit.

**Comment by Nekroze** [comment]
@MartinNowak 
So in that first link does the static this get called at compile time with any contant values (such as a string in my case) in its scope? So without ever executing the block the template is instantiated in the index still gets populated?

**Comment by MartinNowak** [comment]
> @MartinNowak 
> So in that first link does the static this get called at compile time with any contant values (such as a string in my case) in its scope? So without ever executing the block the template is instantiated in the index still gets populated?

It gets instantiated at compile time with the constant string, but it registers that string at runtime.
This is an idiom to collect something from all template instances.

**Comment by Nekroze** [comment]
@MartinNowak 
Beautiful. I will work on incorporating this over the weekend into `DuoD` if I can.

**Comment by Nekroze** [comment]
@MartinNowak 
is there a reason you use `__gshared` instead of `shared` for `registerQuery.fields`?

**Comment by Nekroze** [comment]
@MartinNowak 
Started work on implementing this, as can be seen [here](https://projects.nekroze.com/projects/duod/repository/entry/source/duod/pipeline.d?rev=feature%2Fpipeline).

I am think about the mechanism to build all of the assets. To avoid interfering with any other command line interfaces I may have a `shared static this` block on a module in `DuoD` that checks the command line parameters for a specific switch like `--duod-build`.

**Comment by Nekroze** [comment]
@MartinNowak 
Ok, more work on integrating this idea to bring the asset construction out of unittests.

Just merged in the feature branch I was working on. I realised that I would have to parse the command line args to have a build triggered if the right switch was given without the end user of the library calling a build function in there code. 

I ended up putting this logic in the `static this` block of the template as this way I did not actually need to utilize the idiom you outlined earlier to collect the resources passed to all of the templates. My thinking in using a `static this` block rather then `shared static this` within the template is that I would like to offer the end user a chance to change some of the variables in a `shared static this` block somwhere if they did not use the `public` dir as their static web directory.

I am just about to start writing up an example project or two to prove and demonstrate the functionality.

Thank you very much for your assistance Martin.
